File: WebDriverTests_Java/WebDriverTests/src/test/java/core/systemControls/Waiter.java
package core.systemControls;

import core.supportClass.StopWatch;
import org.junit.Assert;
import org.openqa.selenium.ElementNotVisibleException;
import org.openqa.selenium.WebDriverException;

public final class Waiter
{
    private static final int waitTimeout = 100;
    private static final int defaultTimeout = 30000;
    private static final int firstTestTimeoutFactor = 3;
    private static String firstTestName;


    public static void Wait(Function<Boolean> tryFunc, String actionDescription) {
        Wait(tryFunc, actionDescription, null);
    }

    public static void Wait(Function<Boolean> tryFunc, String actionDescription, Integer timeout) {
        if (timeout==null) timeout=defaultTimeout;
        StopWatch w = new StopWatch();
        w.start();
        do
        {
            if (tryFunc.invoke()) {
                return;
            }
            try {
                Thread.sleep(waitTimeout);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        } while (w.getElapsedTimeSecs()*1000 < GetActualTimeout(timeout));
        w.stop();
        Assert.fail(String.format("действие '%1$s' не выполнилось за '%2$s' мс", actionDescription, GetActualTimeout(timeout)));
    }



    @FunctionalInterface
    public interface Function<Boolean>
    {
        Boolean invoke();
    }
    public Boolean invoke()
    {
        return invoke();
    }

    @FunctionalInterface
    public interface Action
    {
        void invokeVoid();
    }

    public void invokeVoid()
    {
        invokeVoid();
    }



    private static int GetActualTimeout(Integer timeout)
    {
        return (timeout != null) ? timeout : defaultTimeout;
    }

    public static void WaitExeption(Action tryFunc    ) {
        WaitExeption(tryFunc, "WaitExeption", null);
    }
    public static void WaitExeption(Action tryFunc, String actionDescription) {
        WaitExeption(tryFunc, actionDescription, null);
    }

    public static void WaitExeption(Action tryFunc, String actionDescription, Integer timeout) {
        if (timeout==null) timeout=defaultTimeout;
        StopWatch w = new StopWatch();
        w.start();
        do
        {
            try{
                tryFunc.invokeVoid();
                return;
            }
            catch (ElementNotVisibleException e)
            {
                try {
                    Thread.sleep(100);
                } catch (InterruptedException e3) {
                    e3.printStackTrace();
                }
            }
            catch (WebDriverException e)
            {
                try {
                    Thread.sleep(100);
                } catch (InterruptedException e1) {
                    e1.printStackTrace();
                }
            }
        } while (w.getElapsedTimeSecs()*1000 < GetActualTimeout(timeout));
        w.stop();
        Assert.fail(String.format("действие '%1$s' не выполнилось за '%2$s' мс", actionDescription, GetActualTimeout(timeout)));
    }
}

File: WebDriverTests_Csharp/Autotest.Core/Autotests.Utilities/WebTestCore/TestSystem/Waiter.cs
using System;
using System.Diagnostics;
using System.Threading;
using NUnit.Framework;

namespace Autotests.Utilities.WebTestCore.TestSystem
{
    public static class Waiter
    {
        private const int waitTimeout = 100;
        private const int defaultTimeout = 20000;
        private const int firstTestTimeoutFactor = 3;
        private static string firstTestName;

        public static void Wait(Func<bool> tryFunc, string actionDescription, int? timeout = null)
        {
////            if (TeamCityEnvironment.IsExecutionViaTeamCity)
////            {
            if (TestContext.CurrentContext != null && TestContext.CurrentContext.Test != null &&
                TestContext.CurrentContext.Test.FullName != null)
            {
                if (firstTestName == null || firstTestName == TestContext.CurrentContext.Test.FullName)
                {
                    timeout = GetActualTimeout(timeout)*firstTestTimeoutFactor;
                    firstTestName = TestContext.CurrentContext.Test.FullName;
                }
            }
//            }
            DoWait(tryFunc,
                () => Assert.Fail("действие {0} не выполнилось за {1} мс", actionDescription, GetActualTimeout(timeout)),
                timeout);
        }

        public static void DoWait(Func<bool> tryFunc, Action doIfWaitFails, int? timeout = null)
        {
            Stopwatch w = Stopwatch.StartNew();
            do
            {
                if (tryFunc())
                    return;
                Thread.Sleep(waitTimeout);
            } while (w.ElapsedMilliseconds < GetActualTimeout(timeout));
            doIfWaitFails();
        }

        private static int GetActualTimeout(int? timeout)
        {
            return timeout ?? defaultTimeout;
        }
    }
}